src/compute/src/memory/config.rs

```diff
@@ -55,17 +57,44 @@ pub fn reserve_memory_bytes(opts: &ComputeNodeOpts) -> (usize, usize) {
         );
     }
 
-    // If `reserved_memory_bytes` is not set, use `SYSTEM_RESERVED_MEMORY_PROPORTION` * `total_memory_bytes`.
-    let reserved = opts.reserved_memory_bytes.unwrap_or_else(|| {
-        (opts.total_memory_bytes as f64 * SYSTEM_RESERVED_MEMORY_PROPORTION).ceil() as usize
-    });
+    // If `reserved_memory_bytes` is not set, calculate total_memory_bytes based on gradient reserve memory proportion.
+    let reserved = opts
+        .reserved_memory_bytes
+        .unwrap_or_else(|| gradient_reserve_memory_bytes(opts.total_memory_bytes));
 
     // Should have at least `MIN_SYSTEM_RESERVED_MEMORY_MB` for reserved memory.
     let reserved = std::cmp::max(reserved, MIN_SYSTEM_RESERVED_MEMORY_MB << 20);
 
     (reserved, opts.total_memory_bytes - reserved)
 }
 
+/// Calculate the reserved memory based on the total memory size.
+/// The reserved memory size is calculated based on the following gradient:
+/// - 30% of the first 16GB
+/// - 25% of the next 16GB
+/// - 20% of the next 32GB
+/// - 15% of the rest
+fn gradient_reserve_memory_bytes(total_memory_bytes: usize) -> usize {
```

Comment:
What does the `gradient reserve memory proportion` mean? May I know the background for those magic numbers in the `RESERVED_MEMORY_LEVELS` and `RESERVED_MEMORY_PROPORTIONS`?

Reply:
Previously, we used 30% as reserved memory which would consume too much memory in a machine with a large memory capacity. We want to reserve less memory for large machines but keep the original proportion for small machines, so a gradient proportion approach is introduced.

Reply:
IIUC, for a 64GB memory machine, we will reserve 64 * 0.3 = 19.2GB previously. In this PR, we will reserve 16*0.3+16*0.25+32*0.2 = 15.2GB.

Reply:
Yes. The unit test includes a spreadsheet detailing various configurations of reserved memory relative to total memory.